ddtrace/tracer/stats.go

```diff
@@ -99,23 +99,30 @@ func (c *concentrator) Start() {
 	}()
 }
 
+// send sends a stats payload using the transport
+func (c *concentrator) send(sp *statsPayload) {
+	if len(sp.Stats) == 0 {
+		// nothing to flush
+		return
+	}
+	c.statsd().Incr("datadog.tracer.stats.flush_payloads", nil, 1)
+	c.statsd().Incr("datadog.tracer.stats.flush_buckets", nil, float64(len(sp.Stats)))
+	if err := c.cfg.transport.sendStats(sp); err != nil {
+		c.statsd().Incr("datadog.tracer.stats.flush_errors", nil, 1)
+		log.Error("Error sending stats payload: %v", err)
+	}
+}
+
 // runFlusher runs the flushing loop which sends stats to the underlying transport.
 func (c *concentrator) runFlusher(tick <-chan time.Time) {
 	for {
 		select {
 		case now := <-tick:
-			p := c.flush(now)
-			if len(p.Stats) == 0 {
-				// nothing to flush
-				continue
-			}
-			c.statsd().Incr("datadog.tracer.stats.flush_payloads", nil, 1)
-			c.statsd().Incr("datadog.tracer.stats.flush_buckets", nil, float64(len(p.Stats)))
-			if err := c.cfg.transport.sendStats(&p); err != nil {
-				c.statsd().Incr("datadog.tracer.stats.flush_errors", nil, 1)
-				log.Error("Error sending stats payload: %v", err)
-			}
+			sp := c.flush(now, withoutCurrentBucket)
+			c.send(&sp)
 		case <-c.stop:
+			sp := c.flush(time.Now(), withCurrentBucket)
+			c.send(&sp)
```

Comment:
A few questions here:

* Why are you using `now` on the first call, but `time.Now()` on the second call
* Is it necessary for `send` and `flush` to be separate? I liked your initial implementation more.
* If the above answer is "yes", then can we make a separate method `flushAndSend` or something to avoid duplication?

Reply:
> Why are you using now on the first call, but time.Now() on the second call

We don't have a `now` available when stop is handled.

> Is it necessary for send and flush to be separate? I liked your initial implementation more.

https://github.com/DataDog/dd-trace-go/pull/1393#discussion_r927041237 was the context for the change. The lock would guard the transport too. 

If you're ok with this then I am too, I just noticed it.

> If the above answer is "yes", then can we make a separate method flushAndSend or something to avoid duplication?

We could explicitly release the lock rather than deferring to avoid the above if the lock acquisition during transmission is indeed a problem. Else we just keep it as I had it before 🙂